### PR TITLE
Add Video Llava 

### LINKED ISSUE #29640: Video-LLaVA with transformers library
### Model description

Video-LLaVA is a multimodal model that is trained on both images and videos simultaneously.  I feel that it would be highly beneficial if it is added to the transformers library as it is a pretty good choice when it comes to video-question answering

### Open source status

- [ ] The model implementation is available
- [ ] The model weights are available

### Provide useful links for the implementation

https://huggingface.co/LanguageBind/Video-LLaVA-7B

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

Fixes #29640 . Add a new model, Video Llava to the library. 

This is a draft PR, will add more here later.

The model now accepts both, video and image as input in the same batch. Each visual has its own special token, so we do not need to repeat "<image>" 8 times for 8 frames. Here is a short code snippet:  
```python
import torch
import numpy as np
import requests
from PIL import Image
from decord import VideoReader
from huggingface_hub import hf_hub_download
from transformers import VideoLlavaForConditionalGeneration, VideoLlavaProcessor


def load_video_tensor(video_path, n_frms=4, transform=False):
    vr = VideoReader(uri=video_path, height=224, width=224)
    n_frms = min(n_frms, len(vr))
    indices = np.arange(0, len(vr), len(vr) / n_frms).astype(int)
    frames = vr.get_batch(indices).asnumpy()
    return frames

# -------------------------------------------------------------------------------------------------------------------

model = VideoLlavaForConditionalGeneration.from_pretrained("/home/raushan/video_llava/")

url = "http://images.cocodataset.org/val2017/000000039769.jpg"
image = Image.open(requests.get(url, stream=True).raw)

video_path = hf_hub_download(repo_id="raushan-testing-hf/videos-test", filename="sample_demo_1.mp4", repo_type="dataset")
clip = load_video_tensor(video_path, n_frms=8)

processor = VideoLlavaProcessor.from_pretrained("/home/raushan/video_llava")
prompt_vid = "USER: <video>What do you see? ASSISTANT: A baby with a remote control. USER: <video>Why is this funny? ASSISTANT:"
prompt_img = "USER: <image>How many cats are there in the image? ASSISTANT:"
prompt_mix = "USER: <image>How many cats are there in the image? ASSISTANT: 2. USER: <video>Why is this video funny? ASSISTANT:"

inputs = processor(text=[prompt_mix, prompt_img], visual_inputs=[image, clip, image], padding=True, return_tensors="pt")

out = model.generate(**inputs, max_new_tokens=40)
print(processor.batch_decode(out, skip_special_tokens=True, clean_up_tokenization_spaces=True))

# [
#    'USER:  How many cats are there in the image? ASSISTANT: 2. USER:  Why is this video funny? ASSISTANT: The video is funny because the baby is sitting on 
#         the bed and reading a book, which is an unusual and amusing sight. Babies are typically not known for reading books, and the fact', 
#    'USER:  How many cats are there in the image? ASSISTANT: There are two cats in the image..'
# ]


```

### CODE REVIEW

**Comment by zucchini-nlp** [comment]
@LinB203 hey! As we talked before, here is a draft PR of Video Llava. I checked that the modeling part runs without errors and generates similar to the original repo.

To update model files on the hub, you can use `convert_weight` script and use this branch to test if model is loading correctly. Whenever you are available, can you look through, if I missed anything important? :)

**Comment by zucchini-nlp** [comment]
@LinB203 pinging in case the first one got lost in notifications :)

**Comment by ArthurZucker** [comment]
FYI @NielsRogge  and @amyeroberts

File: README.md
Comment (NielsRogge) [inline README.md:526]:
```suggestion
1. **[Video-LLaVA](https://huggingface.co/docs/transformers/main/model_doc/video_llava)** (from YUAN Lab, Peking University) released with the paper [Video-LLaVA: Learning United Visual Representation by Alignment Before Projection](https://arxiv.org/abs/2311.10122) by Bin Lin, Yang Ye, Bin Zhu, Jiaxi Cui, Munang Ning, Peng Jin, Li Yuan.
```
Please link to abstracts instead of pdf :)

File: src/transformers/models/auto/configuration_auto.py
Comment (NielsRogge) [inline src/transformers/models/auto/configuration_auto.py:472]: This can be removed everywhere since this is no longer used

**Comment by zucchini-nlp** [comment]
I believe we can start reviewing this now. I converted weights and added them to my hub account temporarily, so that we can run and test the model. 

In the meanwhile will be waiting for @LinB203 to update the weights in the organization hub, which I think is required before we merge the PR

**Comment by amyeroberts** [comment]
@zucchini-nlp Awesome work! First thing to do before a full review is resolving the failing tests. Some of these are unrelated, and rebasing on main should resolve. Looking at the CI, some of the failure as video llava specific - let me know if you need any help addressing them!

**Comment by zucchini-nlp** [comment]
Rebased with main and resolved conflicts. The only failing doctest seems to be not able to load and run 7b model in 120sec, but I think we will leave it anyway to show how Video-Llava works

**Comment by amyeroberts** [comment]
@zucchini-nlp You can exclude the model from running in the doc tests by adding it to slow_documentation_tests.txt. 

Then, once the PR is reviewed in a steady state and ready for merging, we can run the slow tests and the documentation tests to make sure everything is correct before merging

File: src/transformers/models/video_llava/image_processing_video_llava.py
Comment (amyeroberts) [inline src/transformers/models/video_llava/image_processing_video_llava.py:216]: We might want to define a new type here - as it should be possible to pass in a video as a tensor of size `(batch_size, num_frames, num_channels, height, width)`

```suggestion
        images: ImageInput,
        videos: VideoInput,
```

File: src/transformers/models/video_llava/modeling_video_llava.py
Comment (amyeroberts) [inline src/transformers/models/video_llava/modeling_video_llava.py:130]: This should be handled on the size of the vision towers, rather than this model i.e. we shouldn't know which model is loaded, as we're using `AutoModel`

File: src/transformers/models/video_llava/modeling_video_llava.py
Comment (amyeroberts) [inline src/transformers/models/video_llava/modeling_video_llava.py:162]: ~This isn't going to be compatible if we have vision tower's with attention which don't support sdpa. ~

I see below it's handled by passing the attn implementation just to the text tower. I think this is bit confusing  as this value only applies to the text tower and not the whole model.

File: src/transformers/models/video_llava/modeling_video_llava.py
Comment (amyeroberts) [inline src/transformers/models/video_llava/modeling_video_llava.py:165]: Missing inputs in docstring

File: src/transformers/models/video_llava/modeling_video_llava.py
Comment (amyeroberts) [inline src/transformers/models/video_llava/modeling_video_llava.py:434]: We should use av instead of decord. decord causes crashes on certain GPU setups and isn't compatible with windows

File: src/transformers/models/video_llava/modeling_video_llava.py
Comment (amyeroberts) [inline src/transformers/models/video_llava/modeling_video_llava.py:406]: It's not ideal having to change the standard input type names. I _think_ this is OK, as we have e.g. `input_ids_query`, `input_ids_doc` for retribert

File: src/transformers/models/video_llava/modeling_video_llava.py
Comment (amyeroberts) [inline src/transformers/models/video_llava/modeling_video_llava.py:290]: Is there only ever 1 image token is there's an image i.e. for image inputs `max_seq_len == sequence_length`

File: tests/models/video_llava/test_modeling_video_llava.py
Comment (amyeroberts) [inline tests/models/video_llava/test_modeling_video_llava.py:376]: 'What is this generation funny: because the addition of "hopefully" makes it seem weirdly sinister' 👀

File: tests/models/video_llava/test_modeling_video_llava.py
Comment (amyeroberts) [inline tests/models/video_llava/test_modeling_video_llava.py:404]: Are we expecting the double space after USER here?

File: tests/models/video_llava/test_modeling_video_llava.py
Comment (amyeroberts) [inline tests/models/video_llava/test_modeling_video_llava.py:563]: lets just force the formatting off here to make it nicer to view 

```suggestion
	# fmt: off
        input_ids = torch.tensor(
            [
                [
                    32001, 32001, 1, 15043, 7084, 32000, 32000, 32000, 32000, 32000, 32000, 32000, 32000, 29871, 13, 7900
                ],
                [
                    1, 15043, 7084, 29901, 29871, 32000, 32000, 32000, 32000, 32000, 32000, 32000, 32000, 29871, 13, 7900
                ],
            ],
            dtype=torch.long,
            device=torch_device,
        )
        # fmt: on
```

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding this model. Super clean PR - was a pleasure to review! 

Mostly just small comments

File: src/transformers/models/video_llava/modeling_video_llava.py
Comment (nileshkokane01) [inline src/transformers/models/video_llava/modeling_video_llava.py:449]: what is 'clip' here?

File: tests/models/video_llava/test_modeling_video_llava.py
Comment (zucchini-nlp) [reply tests/models/video_llava/test_modeling_video_llava.py:404]: yes, the same is in Llava models when we add the special "<image>" token

File: src/transformers/models/video_llava/modeling_video_llava.py
Comment (zucchini-nlp) [reply src/transformers/models/video_llava/modeling_video_llava.py:290]: If the prompt is only "< image >" it also works without errors. Also I added tests for inputs with interleaved and image/video only prompts.

File: src/transformers/models/video_llava/modeling_video_llava.py
Comment (zucchini-nlp) [reply src/transformers/models/video_llava/modeling_video_llava.py:449]: will be fixed, should be the `frames`

File: src/transformers/models/video_llava/configuration_video_llava.py
Comment (zucchini-nlp) [reply src/transformers/models/video_llava/configuration_video_llava.py:123]: Will remove. That was copied from one of the old Llava versions, but I see that `vocab_size` for VLMs is deprecated now

File: src/transformers/models/video_llava/processing_video_llava.py
Comment (zucchini-nlp) [reply src/transformers/models/video_llava/processing_video_llava.py:53]: Okay, now it can be removed. It was needed before I added special token for video

File: src/transformers/models/video_llava/modeling_video_llava.py
Comment (zucchini-nlp) [reply src/transformers/models/video_llava/modeling_video_llava.py:524]: Yes, VideoLlava was trained and has to be used with 8 video frames. I will add it in the model docs page in "usage tips" section

**Review by amyeroberts — COMMENTED** [review]
Thanks for the continued work on this!

File: src/transformers/models/video_llava/configuration_video_llava.py
Comment (zucchini-nlp) [reply src/transformers/models/video_llava/configuration_video_llava.py]: I haven't seen a condition when it as missing and it's actually a copy from Llava-based models configuration file. I will raise  a warning to be explicit here for users, I guess default-clip comes from the fact that llavas are based on clip.

File: src/transformers/models/video_llava/configuration_video_llava.py
Comment (zucchini-nlp) [reply src/transformers/models/video_llava/configuration_video_llava.py:98]: The `self.vision_feature_select_strategy` is checked a few lines above, in `self._get_vision_features()`. We try to get the feature and if not raise ValueError

File: tests/test_modeling_common.py
Comment (zucchini-nlp) [reply tests/test_modeling_common.py:1765]: okay :)

File: src/transformers/models/video_llava/processing_video_llava.py
Comment (zucchini-nlp) [reply src/transformers/models/video_llava/processing_video_llava.py:118]: Hmm we cannot have no text, should be at least "< image >" to indicate that there is an image. I will remove the `text is None` locally and commit suggestion for now

File: tests/models/video_llava/test_modeling_video_llava.py
Comment (zucchini-nlp) [reply tests/models/video_llava/test_modeling_video_llava.py:41]: removed, was a copy from llava

File: src/transformers/models/video_llava/image_processing_video_llava.py
Comment (zucchini-nlp) [reply src/transformers/models/video_llava/image_processing_video_llava.py:1]: Added tests, but there is one thing to note. If we call directly the ImageProcessor class, it requires and argument `images` to be present. A workaround is to pass explicitly `images=None` for VideoLlavaImageProcessor, which I did for the tests.

I can override _call_ and to make the argument `images = None`. so that it is optional, but not sure how good is overriding _call_. Also, I do not think many ppl call image processor explicitly.

Comment (amyeroberts) [reply src/transformers/models/video_llava/image_processing_video_llava.py:1]: If the image processor takes both `images` and `videos` as input, and only one of them is required, then setting `image = None` seems reasonable

File: src/transformers/models/video_llava/modeling_video_llava.py
Comment (amyeroberts) [reply src/transformers/models/video_llava/modeling_video_llava.py:550]: Has this been added? Skimming I didn't spot but might have just missed

**Review by amyeroberts — APPROVED** [review]
Thanks for all the work adding this model - looks great!

File: src/transformers/models/video_llava/modeling_video_llava.py
Comment (zucchini-nlp) [reply src/transformers/models/video_llava/modeling_video_llava.py:550]: It was added in `_get_vision_features()`, after that we can never know how many frames we have

**Comment by zucchini-nlp** [comment]
Hey @LinB203, can you let us know if you can upload HF weights of VideoLlava to the organization? The model seems rready to be added to the library

**Comment by LinB203** [comment]
> Hey @LinB203, can you let us know if you can upload HF weights of VideoLlava to the organization? The model seems rready to be added to the library

Thanks for your great attention and employing the model in transformers. I wonder that which organization you mentioned?

**Comment by zucchini-nlp** [comment]
@LinB203 I mean the weights that can loaded into the `transformers` model that were converted by the "convert_weights" script I added in this PR. I already have it tested and added weights to my account in the hub, but it would be nice if it was under the "LanguageBind" org. If you need to keep the current version of the state dict, you can call the new model as "LanguageBind/Video-LLaVA-7B-hf" prob

**Comment by LinB203** [comment]
> @LinB203 I mean the weights that can loaded into the `transformers` model that were converted by the "convert_weights" script I added in this PR. I already have it tested and added weights to my account in the hub, but it would be nice if it was under the "LanguageBind" org. If you need to keep the current version of the state dict, you can call the new model as "LanguageBind/Video-LLaVA-7B-hf" prob

I see. I will make `LanguageBind/Video-LLaVA-7B-hf` repo and upload new converted weight by your script tonight. Thank you very much.

**Comment by LinB203** [comment]
> @LinB203 I mean the weights that can loaded into the `transformers` model that were converted by the "convert_weights" script I added in this PR. I already have it tested and added weights to my account in the hub, but it would be nice if it was under the "LanguageBind" org. If you need to keep the current version of the state dict, you can call the new model as "LanguageBind/Video-LLaVA-7B-hf" prob

Finished.  https://huggingface.co/LanguageBind/Video-LLaVA-7B-hf/tree/main

**Comment by zucchini-nlp** [comment]
@amyeroberts done! The last commit should trigger the slow tests after your approval. 

Note about the failing code-check, it's failing because of the `ARCHIVE_MAP` which apparently was removed for all models. So I didn't add it for VideoLlava

**Comment by amyeroberts** [comment]
@zucchini-nlp Great! :D Could you rebase to include the upstream changes like the `ARCHIVE_MAP` removal? This should make everything green and ensure it's just that triggering the errors

**Comment by zucchini-nlp** [comment]
The PR passed all the tests, slow tests are passing for me locally. Should be good to go

**Comment by zucchini-nlp** [comment]
Just fixed one slow test, will merge when i get all green

**Pull request merged**
